torch/ao/quantization/pt2e/numeric_debugger.py

```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
+    """For a given node, adds an OutputLogger that observes the output of that node,
+    and all its users use the OutputLogger output instead.
+    The OutputLogger will contain the debug_handle which can be used to compare
+    graphs after transforms"""
+
+    # to avoid circular dep
+    from torch.ao.quantization.fx.utils import get_new_attr_name_with_prefix
+
+    # add a logger after the node
+    with model.graph.inserting_after(node):
+        get_new_attr_name = get_new_attr_name_with_prefix(f"{node.name}_logger")
+        logger_name = get_new_attr_name(model)
+        setattr(
+            model,
+            logger_name,
+            OutputLogger(node.name, node.meta.get("nn_module_stack"), debug_handle),
+        )
+        logger_node = model.graph.call_module(logger_name, (node,), {})
+
+    orig_users = list(node.users.keys())
+    for user_node in orig_users:
+        if user_node is logger_node:
+            continue
+        user_node.replace_input_with(node, logger_node)
+
+
+def prepare_for_propagation_comparison(model: GraphModule) -> GraphModule:
+    # don't change the original model
+    model = copy.deepcopy(model)
+    for n in model.graph.nodes:
+        if NUMERIC_DEBUG_HANDLE_KEY not in n.meta:
+            continue
+        numeric_debug_handle = n.meta[NUMERIC_DEBUG_HANDLE_KEY]
+        _insert_logger(model, n, numeric_debug_handle)
+
+    model.recompile()
+    return model
+
+@dataclass(frozen=True)
+class QuantizationComparisonResult:
+    actual: torch.Tensor
+    ref: torch.Tensor
+
+    @property
+    def mse_loss(self) -> torch.Tensor:
+        return F.mse_loss(
+            self.actual.to(dtype=torch.float32), self.ref.to(dtype=torch.float32)
+        )
+
+    @property
+    def sqnr(self) -> torch.Tensor:
+        return compute_sqnr(
+            self.actual.to(dtype=torch.float32), self.ref.to(dtype=torch.float32)
+        )
+
+    def details(self) -> str:
+        return f"QuantizationComparisonResult(mse_loss={self.mse_loss}, sqnr={self.sqnr}, actual={self.actual}, ref={self.ref})"
+
+    def __repr__(self) -> str:
+        # Don't include the tensors themselves as they are quite large to print
+        # out.
+        return (
+            f"QuantizationComparisonResult(mse_loss={self.mse_loss}, sqnr={self.sqnr})"
+        )
+
+
+@dataclass(frozen=True)
+class NodeAccuracySummary:
+    handle: str
+    actual_node_name: str
+    actual_module_stack: str
+    ref_node_name: str
+    ref_module_stack: str
+    results: Sequence[QuantizationComparisonResult]
+
+    def details(self) -> str:
+        """Returns a very detailed string for the results including all tensors.
+        If you don't want to see this, just use str() or repr() on this object instead
+        to get a more concise summary."""
+        s = f"{repr(self)}\ndetails:\n"
+        for q in self.results:
+            s += q.details() + "\n"
+        return s
+
+
+def _module_stack_to_str(module_stack: object) -> str:
```

Comment:
You might want to put this function somewhere else, I'm sure other people like to visualize the nn_module_stack

Reply:
this seems to be a small util, we can move when other people also needs this I think